federation/pkg/federation-controller/namespace/namespace_controller.go

```diff
@@ -0,0 +1,353 @@
+			return framework.NewInformer(
+				&cache.ListWatch{
+					ListFunc: func(options api.ListOptions) (pkg_runtime.Object, error) {
+						return targetClient.Core().Namespaces().List(options)
+					},
+					WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+						return targetClient.Core().Namespaces().Watch(options)
+					},
+				},
+				&api.Namespace{},
+				controller.NoResyncPeriodFunc(),
+				framework.ResourceEventHandlerFuncs{
+					// Trigger reconcilation whenever something in federated cluster is changed. In most cases it
+					// would be just confirmation that some namespace opration suceeded.
+					DeleteFunc: func(old interface{}) {
+						// Remove once it is done automatically.
+						nold, _ := old.(*api_v1.Namespace)
+						util.SetClusterName(nold, cluster.Name)
+						nc.deliverNamespaceObj(old, NamespaceReviewDelay, 0)
+					},
+					AddFunc: func(cur interface{}) {
+						// Remove once it is done automatically.
+						ncur, _ := cur.(*api_v1.Namespace)
+						util.SetClusterName(ncur, cluster.Name)
+						nc.deliverNamespaceObj(cur, NamespaceReviewDelay, 0)
+					},
+					UpdateFunc: func(old, cur interface{}) {
+						nold, _ := old.(*api_v1.Namespace)
+						ncur, _ := cur.(*api_v1.Namespace)
+						// Remove once it is done automatically.
+						util.SetClusterName(nold, cluster.Name)
+						util.SetClusterName(ncur, cluster.Name)
+
+						if !reflect.DeepEqual(nold.Spec, ncur.Spec) ||
+							!reflect.DeepEqual(nold.ObjectMeta, ncur.ObjectMeta) {
+							nc.deliverNamespaceObj(cur, NamespaceReviewDelay, 0)
+						}
+					},
+				})
+		},
+		&util.ClusterLifecycleHandlerFuncs{
+			ClusterAvailable: func(cluster *federation_api.Cluster) {
+				// When new cluster becomes available process all the namespaces again.
+				nc.clusterDeliverer.DeliverAt(allClustersKey, nil, time.Now().Add(ClusterAvailableDelay))
+			},
+		},
+	)
+
+	// Federated updeater along with Create/Update/Delete operations.
+	nc.federatedUpdater = util.NewFederatedUpdater(nc.namespaceFederatedInformer,
+		func(client federation_release_1_4.Interface, obj pkg_runtime.Object) error {
+			namespace, _ := obj.(*api_v1.Namespace)
+			_, err := client.Core().Namespaces().Create(namespace)
+			return err
+		},
+		func(client federation_release_1_4.Interface, obj pkg_runtime.Object) error {
+			namespace, _ := obj.(*api_v1.Namespace)
+			_, err := client.Core().Namespaces().Update(namespace)
+			return err
+		},
+		func(client federation_release_1_4.Interface, obj pkg_runtime.Object) error {
+			namespace, _ := obj.(*api_v1.Namespace)
+			err := client.Core().Namespaces().Delete(namespace.Name, &api.DeleteOptions{})
+			return err
+		})
+	return nc
+}
+
+func (nc *NamespaceController) Start() {
+	nc.namespaceInformerController.Run(nc.namespaceStopChan)
+	nc.namespaceFederatedInformer.Start()
+
+	// Handle cluster available/unavailable events.
+	go func() {
+		for {
```

Comment:
Please split this to a separate function and just call go thisFunction() here.

Same below.